L46-core-tls-credential-API.md

```diff
@@ -0,0 +1,369 @@
+L46: C-core: New TLS Credentials API
+----
+* Author(s): gtcooke94, ZhenLian
+* Approver: 
+* Status: Draft
+* Implemented in:
+* Last updated: May 16, 2024
+
+## Abstract
+
+This proposal aims to provide the following features in a new TLS credential API:
+1. credential reloading: reload transport credentials periodically from various sources, e.g. from credential files, or users' customized Provider implementations, without shutting down
+1. custom verification behaviors: performing a user-specified check at the end of normal authentication and separately giving the ability to fully customize chain building and verification.
+1. TLS version configuration: optionally configure the minimum and maximum of the TLS version that will be used in the handshake
+1. support for certificate revocation via Certificate Revocation Lists (CRLs)
+
+## Background
+
+The current TLS implementation in gRPC C core has some restrictions:
+1. lack of the credential reloading support. Once an application starts, it can never reload its identity key/cert-chain and root certificates without shutting down
+1. inflexibility in the peer verification. In a mutual TLS scenario, servers would always perform certificate verification, and clients would always perform certificate verification as well as default hostname check. No customized checks can be introduced.
+1. inability to choose the maximum/minimum TLS version supported in OpenSSL/BoringSSL for use
+
+There have always been some ad hoc requests on GitHub Issue Page for supporting these features. The demands for a more flexible and configurable API have increased with the emerging needs in the cloud platform.
+
+One common requirement is to skip the hostname verification check if a client has a different way of validating the server's identity in different platforms. 
+For example, it's becoming more and more popular to use SPIFFE ID as the identity format, but a valid certificate using SPIFFE ID doesn't necessarily contain the dnsName in the SAN field, which might cause hostname verification to fail.
+
+Another requirement is to reload the root and identity credentials without shutting down the server, because certificates will eventually expire and the ability to rotate the certificates is essential to a cloud application.
+
+Besides requests from Open Source Community, with the growth of GCP, we've seen more and more internal use cases of these features.   
+
+Hence we propose an API that will meet the following requirements:  
+
+1. Flexible enough to support multiple use cases. Users should be able to write their own reloading or verification logic if the provided ones doesn't fit their needs
```

Comment:
micro-nit: Missing period at the end of the sentence, similarly in the next bullet points.